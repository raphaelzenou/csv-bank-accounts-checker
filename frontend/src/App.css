.app {
  text-align: center;
}

.app-logo {
  height: 40vmin;
  pointer-events: none;
}

.app-header {
  background-color: #1f2f52;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}
.app-instructions {
  padding: 1em 0;
}

pre {
  color: #ffe534;
  font-size: 0.75em;
}

.csv-reader-input__button {
  visibility: hidden;
  width: 0;
  height: 0;
}

.process-button, .csv-reader-input__label {
  color: #1f2f52;
  background-color: #fff;
  padding: 0.2em;
  font-size: 1em;
  border-radius: 0.3em;
  border: none;
  margin: 1em 0;
  outline: 0;
  cursor: pointer;
}


.process-button:disabled, .csv-reader-input__label:disabled {
  opacity: 50%;
  cursor: not-allowed;
}