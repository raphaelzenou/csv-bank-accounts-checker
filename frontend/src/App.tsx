import React, { useState } from 'react';
import CSVReader from "react-csv-reader";
import './App.css';
import Logo from './logo.png';

/**
 * Given the size of this app
 * Choice has been made to keep 
 * all functionality in the same module
 */

function App() {

  // local state
  const [csvData, setCsvData] = useState<any[] | [] >([]);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [isOkToProcess, setIsOkToProcess] = useState<boolean>(false);
  const [message, setMessage ] = useState<string | Promise<string> >('Please upload a trial balance CSV file to check its format.');

  interface IFileInfo {
    name: string,
    size: number,
    type: string,
  }

  interface IResponse {
    status: number,
    ok: boolean,
    message: string,
  }
  // re-initialising state
  const resetApp = () : void  => {
    // local state
    setCsvData([]);
    setIsOkToProcess(false);
    setIsLoading(false);
    setMessage('Please upload a trial balance CSV file to check its format.');

    // below line is necessary to get the onFileLoaded
    // prop function of CSVReader to work with repeated
    // uploads of the same file

    /** TODO - replace with useRef */
    (document.querySelector('.csv-reader-input__button') as HTMLInputElement).value = '';
  }
  // action on csv file upload
  const onFileUpload = (data : any[], fileInfo : IFileInfo) : void => {
    console.table('DATA', data);

    if (data.length > 0) {
      if(fileInfo?.type === 'text/csv'){ 
        setCsvData(data);
        setIsOkToProcess(true);
        setMessage(`File "${fileInfo.name}" is ready to get processed.`);
      } else {
        setIsOkToProcess(false);
      }
    } else {
      setIsOkToProcess(false);
      setMessage('The CSV file provided is empty.');
    }
  }

  // action on press of the Process button
  const onClickProcess = async () : Promise<void> => {
    setIsOkToProcess(false);
    setIsLoading(true);
    setMessage('Loading...')

    try {

      const response: Response  = await postData('/api', csvData);
      const json : IResponse = await response.json();
      setIsLoading(false);
      setMessage(json.message)

    } catch (error : any) {
      setIsLoading(false);
    } finally {
      setTimeout(()=> resetApp(), 4000);
    }
  }

  // Http POST request helper
  const  postData = async (url: string = '', data: any[] = []) : Promise<Response> => {

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    });
    return response;
  }

  return (
    <div className="app">
      <header className="app-header">
        <img src={Logo} className="app-logo" alt="logo" />
    
        <p>{message}</p>

        <div>
          <CSVReader
            label={'Upload'}
            parserOptions={{
              header: true,
              dynamicTyping: true,
              skipEmptyLines: true,
            }}
            onFileLoaded={(data , fileInfo) => onFileUpload(data, fileInfo)} 
            cssInputClass={'csv-reader-input__button'}
            cssLabelClass={'csv-reader-input__label'}	
          />
        </div>
        <div>
          <button 
            disabled={!isOkToProcess || isLoading} 
            onClick={()=>onClickProcess()} 
            className="process-button">
              Process
          </button>
        </div>
      </header>
    </div>
  );
}

export default App;
