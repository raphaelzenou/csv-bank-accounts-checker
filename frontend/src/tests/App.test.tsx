import React from 'react';
import { render, screen } from '@testing-library/react';
import App from '../App';

it('should render app with the right buttons enabled / disabled', () => {
  const container = render(<App />);

  const uploadButton = screen.getByText('Upload');
  expect(uploadButton).toBeInTheDocument();
  expect(uploadButton).not.toBeDisabled();

  const processButton = screen.getByText('Process');
  expect(processButton).toBeInTheDocument();
  expect(processButton).toBeDisabled();

  expect(container).toMatchSnapshot();

});
