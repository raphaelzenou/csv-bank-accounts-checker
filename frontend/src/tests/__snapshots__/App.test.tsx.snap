// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render app with the right buttons enabled / disabled 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app"
      >
        <header
          class="app-header"
        >
          <img
            alt="logo"
            class="app-logo"
            src="logo.png"
          />
          <p>
            Please upload a trial balance CSV file to check its format.
          </p>
          <div>
            <div
              class="csv-reader-input"
            >
              <label
                class="csv-reader-input__label"
                for="react-csv-reader-input"
              >
                Upload
              </label>
              <input
                accept=".csv, text/csv"
                class="csv-reader-input__button"
                id="react-csv-reader-input"
                type="file"
              />
            </div>
          </div>
          <div>
            <button
              class="process-button"
              disabled=""
            >
              Process
            </button>
          </div>
        </header>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app"
    >
      <header
        class="app-header"
      >
        <img
          alt="logo"
          class="app-logo"
          src="logo.png"
        />
        <p>
          Please upload a trial balance CSV file to check its format.
        </p>
        <div>
          <div
            class="csv-reader-input"
          >
            <label
              class="csv-reader-input__label"
              for="react-csv-reader-input"
            >
              Upload
            </label>
            <input
              accept=".csv, text/csv"
              class="csv-reader-input__button"
              id="react-csv-reader-input"
              type="file"
            />
          </div>
        </div>
        <div>
          <button
            class="process-button"
            disabled=""
          >
            Process
          </button>
        </div>
      </header>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
